### ISSUE TITLE
webrtc Datachannels max Bytes Firefox 51.0.1 -> chrome 56.0.2924.76 

### ISSUE THREAD

**Issue by arputcu** [webrtcHacks/adapter#426]
**Browsers and versions affected**
Firefox 51.0.1 -> chrome 56.0.2924.76 

**Description**
tranfer  data chunks >= 64000K from firefox -> chrome 

**Steps to reproduce**

**Expected results**
transfer 64000k chunks 
works on 
chrome <-> chrome 
chrome -> firefox 

**Actual results**
lose data 

anyone an idea what the problem is? firefox bug or chrome problem?

**Comment by nils-ohlmeier** [comment]
See https://bugzil.la/979417 and https://bugs.chromium.org/p/webrtc/issues/detail?id=2279#c30 and https://webrtc.org/web-apis/chrome/

**Comment by arputcu** [comment]
thx for all this information 
so i  should chunk all my data to <= 16000 chunks on all transfers?

**Comment by KaptenJansson** [comment]
any luck ?

**Comment by arputcu** [comment]
i do not understand the problem:/  it works for 
chrome <-> chrome
ff <-> ff
chrome -> firefox

but not for  ff -> chrome  with the 64k chunks 


and with 16k chunks the transfer performance is not as good as with 64k

**Comment by lgrahl** [comment]
@arputcu Always limit large messages to 16 KiB in interop mode. Check out [my blog posting](https://lgrahl.de/articles/demystifying-webrtc-dc-size-limit.html) for a technical explanation. You can use https://github.com/saltyrtc/chunked-dc-js for user-level fragmentation and reassembly.

---

This adapter should probably...
1. add the [`maxMessageSize` attribute](https://www.w3.org/TR/webrtc/#dom-rtcsctptransport-maxmessagesize), and
2. ensure that the message provided in `RTCDataChannel.send` does not exceed the size indicated by the remote peer (which is in the SDP in form of `a=max-message-size` or, if not present, defaults to 64 KiB).

Behaviour of the API is still open for discussion, see https://github.com/w3c/webrtc-pc/issues/1205

**Comment by nils-ohlmeier** [comment]
@lgrahl I like the idea of adapter taking care of this interop problem by doing the user-level fragmentation and reassembly.
But I'm not sure if or how maxMessageSize would help here. MaxMessageSize is a value each side advertises in it's SDP (not something both sides need to agree on). So Chrome would include a=max-message-size:64000 and Firefox would include a=max-message-size:16000. But then that would still result in the problematic scenario for Firefox trying to send 64KiB message to Chrome. So to prevent that even Chrome would need to advertise a=max-message-size:16000. But at that point this would simply become a hard coded 16KiB for all browsers to ensure interoperability.

Or do I miss something here?

**Comment by lgrahl** [comment]
@nils-ohlmeier Ah, sorry, this is probably a misunderstanding. I don't suggest that adapter takes care of fragmentation and reassembly. This definitely needs to be done in the data channel implementations. The first sentence was just a suggestion for @arputcu. I edited my posting slightly to make this more clear.

> But I'm not sure if or how maxMessageSize would help here. MaxMessageSize is a value each side advertises in it's SDP (not something both sides need to agree on). So Chrome would include a=max-message-size:64000 and Firefox would include a=max-message-size:16000. But then that would still result in the problematic scenario for Firefox trying to send 64KiB message to Chrome.

The whole situation is pretty messy. What I suggested was simply to add some parts of the WebRTC API already described but which no browser implements so far. But you're right, it doesn't make sense to do this at the moment as this only works if the sending side does a) not use PPID-based fragmentation/reassembly, and b) ensures that either arbitrary messages can be sent or raise an exception (or whatever error indication we conclude in w3c/webrtc-pc#1205) in case the message is too large to be sent. Only with all these issues resolved, the two suggestions I made in my previous posting will be useful.
We will get rid of *a)* soon by adding proper EOR-based reassembly to Firefox but *b)* requires addressing from both Firefox and Chrome.
